### PR TITLE
WPCOM: Migrate `wpcom.undocumented()` contact details to `wpcom.req`

### PR DESCRIPTION
**Pull request by tyxla**
#### Changes proposed in this Pull Request

This PR migrates the `wpcom.undocumented()` contact details methods to `wpcom.req`. 

As part of the work, we're reusing the existing `@automattic/wpcom-checkout` package `mapRecordKeysRecursively` that's a substitute of `mapKeysRecursively` and `camelToSnakeCase` that's our custom version of Lodash's `camelCase`. We also introduce a simple `snakeToCamelCase` to serve as our custom version of Lodash's `snakeCase` - cc @sirbrillig on these bits.

We also remove the `_sendRequest` method and the `mapKeysRecursively` helper function that now end up being unused.

Part of the ongoing effort to get rid of `wpcom.undocumented()`.

#### Testing instructions

* Go to `/checkout/somedomain.fr` where `somedomain.fr` is a custom `.fr` domain you're going to purchase. 
* Verify your contact details are retrieved and populated into the form correctly.
* Go to `/domains/manage/:domain/edit-contact-info/:site` where `:domain` and `:site` correspond to one of your sites on a custom domain.
* Play with the validation of the form and verify it still works correctly.
* Verify all tests pass, specifically `yarn run test-packages packages/wpcom-checkout`.
* @Automattic/nomado is there any other flow / screen these need to be tested on?

### CODE REVIEW

File: client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx
Comment (jsnajdr) [inline client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: I'm not very happy about `domains` importing an unrelated package like `@automattic/wpcom-checkout`. Could the functions be in `js-utils`?

Also, I suspect that a generic `mapRecordKeysRecursively` might not be needed. In the POST data, we can specify them as `{ contact_information, domain_names }` directly, can't we?

And what remains is always the same data structure, `contact_information`. That one is flat, with no nested fields, and we could have a custom `fromApi`/`toApi` function that enumerates the fields explicitely. We already have many conversion helpers like this.

Comment (tyxla) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: > I'm not very happy about domains importing an unrelated package like @automattic/wpcom-checkout. Could the functions be in js-utils?

This has been brought up before for other methods, so I think it makes sense. I could spin up a PR for that.

> Also, I suspect that a generic mapRecordKeysRecursively might not be needed. In the POST data, we can specify them as { contact_information, domain_names } directly, can't we?

Yes, however, `contact_information` contains subfields and their keys need to be recursively mapped when retrieving and before sending to the relevant endpoints.

Comment (tyxla) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: > I'm not very happy about domains importing an unrelated package like @automattic/wpcom-checkout. Could the functions be in js-utils?

Given that we need `mapRecordKeysRecursively`, I wonder if you could say the same for the `mapRecordKeysRecursively` method - would you expect to see it in `@automatic/js-utils` as well, @jsnajdr?

Comment (jsnajdr) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: > I wonder if you could say the same for the `mapRecordKeysRecursively` method

At this moment, I'm still convinced we don't need it 🙂 

> however, `contact_information` contains subfields and their keys need to be recursively mapped

Do they really? I tried editing contact info for one of my domains and the `contact_information` structure is flat. Set of fields where the values are always strings.

Comment (tyxla) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: > Do they really? I tried editing contact info for one of my domains and the contact_information structure is flat.

It's not always flat. Try with a `.ca` or `.fr` or `.uk` domain. You'll see nested fields for those:



I can confirm that for any other TLD I've tried, we don't have nested fields.

Comment (jsnajdr) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: I see, there's the `extra` field 🙍‍♂️ Still, the conversion is done only for the "contact information" structure. Then it would be more acceptable, even though not ideal, if the `@automattic/wpcom-checkout` exported functions specific for that:
- `contactInformationFromREST()`
- `contactInformationToREST()`
Implemented internally as some recursive `Object.fromEntries` stuff.

Comment (tyxla) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: Considering that the PR is not introducing `mapRecordKeysRecursively` and it pre-exists, are these contact-specific functions something you insist to see in this PR, or is it something we can get to incrementally? After all, the purpose of this PR is to reuse as much as possible from what we have already in order to migrate away from `wpcom.undocumented()`, not to introduce new APIs or improve existing ones. 

I'd also love to hear @sirbrillig's thoughts about this.

Comment (tyxla) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: It's worth noting that `mapRecordKeysRecursively` and those `camelCase` / `snakeCase` functions are all pretty generic and might make sense to belong to a more generic package like `@automattic/js-utils`. Won't that resolve your concerns, @jsnajdr?

Comment (jsnajdr) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: The only thing that I think should be changed in this PR are the POST data in the validate requests:

before:
```js
.post(
  '/me/domain-contact-information/validate',
  mapRecordKeysRecursively( {
    contactInformation: fieldValues,
    domainNames: [ this.props.selectedDomain.name ],
  }, camelToSnakeCase )
)
```
after:
```js
.post( '/me/domain-contact-information/validate', {
  contact_information: mapRecordKeysRecursively( fieldValues, camelToSnakeCase ),
  domain_names: [ this.props.selectedDomain.name ],
} )
```
That should not be controversial.

The rest is up for discussion. Which structures are we really mapping with these functions? Is it really always "contact information", or its fields, like when the response contains validation erros for fields like `messages: { address_1: 'Enter your address.' }`?

Comment (jsnajdr) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: > Won't that resolve your concerns?

Yes, there's a family of Lodash-inspired object-mapping functions, `mapKeys`, `mapValues`, and `mapKeysRecursively`, that we use a lot and will need to migrate one day. `js-utils` is a good place for them.

`camelCase` and `snakeCase` are another set of helpers that will need to be migrated away from Lodash.

Comment (tyxla) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: > The only thing that I think should be changed in this PR are the POST data in the validate requests:

Oh, I think I've been misunderstanding you. That makes sense, will update it in the next iteration.

>  Which structures are we really mapping with these functions?

It's worth noting that this PR isn't set to answer that kind of questions at all. But we're using generic functions and it's pretty straightforward to trace what we're mapping and how it ends up mapped IMHO.

> Yes, there's a family of Lodash-inspired object-mapping functions

Is this something we can follow-up with? I'm happy to move all those after we merge this PR. I don't think the move necessarily has to block this PR. What do you think?

Comment (jsnajdr) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: > What do you think?

I think the current situation with the mapping functions is not ideal, but we don't need to fix it all in one PR. This `wpcom.undocumented` migration is ready to go, the rest can be independent follow-ups.

File: packages/wpcom-checkout/src/snake-to-camel-case.ts
Comment (sirbrillig) [inline packages/wpcom-checkout/src/snake-to-camel-case.ts]: Did you mean `camelCase`?

File: client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx
Comment (sirbrillig) [inline client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: FWIW, we don't necessarily need to map these as we [already have a typed conversion function](https://github.com/Automattic/wp-calypso/blob/cdce28a85cd69a0318e7d7d166b554487bfa8735/client/my-sites/checkout/composite-checkout/lib/contact-validation.tsx#L213) which is used for the [validation done in checkout](https://github.com/Automattic/wp-calypso/blob/c3fb2139b727879145fa4f747718bb71bed43fd5/client/my-sites/checkout/composite-checkout/lib/contact-validation.tsx#L302). In theory we should be able to re-use that conversion function here although I don't believe I've ever worked on this form so I'm not sure if its input data is in the right format.

File: client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx
Comment (sirbrillig) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: We do know the [expected format of the contact details](https://github.com/Automattic/wp-calypso/blob/c3fb2139b727879145fa4f747718bb71bed43fd5/packages/wpcom-checkout/src/types.ts#L388) and we have [a function that can create it explicitly](https://github.com/Automattic/wp-calypso/blob/c3fb2139b727879145fa4f747718bb71bed43fd5/client/my-sites/checkout/composite-checkout/types/wpcom-store-state.ts#L505) from the camelCase we use in checkout, but I didn't use it when migrating away from `undocumented` for the transactions endpoint because we haven't yet typed the other data sent to the transactions endpoint. Eventually I'd like to get rid of the case changes and mapping in checkout entirely. In theory, those functions could work well here, though.

The biggest problem is probably that the domain contact forms here still use `formStateController` and its input/output is probably not compatible with the conversion functions I mentioned. We migrated away from that class in checkout but as these forms are not checkout related we never touched them. Still, there is a precedent for that work and it could be done as a separate project.

File: client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx
Comment (sirbrillig) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: Hm. I think it might require some data massaging because it looks like these forms still use `formStateController` to manage their data.

File: packages/wpcom-checkout/src/snake-to-camel-case.ts
Comment (tyxla) [reply packages/wpcom-checkout/src/snake-to-camel-case.ts]: Yes! Fixed in https://github.com/Automattic/wp-calypso/pull/58054/commits/b0023a5c367b2a17be4ef6b20b6b432bb16a1318

File: client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx
Comment (tyxla) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: Thank you both!

**Comment by tyxla** [comment]
@jsnajdr I've now moved the `mapRecordKeysRecursively` call to be wrapped around the contact data only.

I think this one is ready for another review as it is. 

In a follow-up PR we could move the `mapRecordKeysRecursively` and the camelcase / snakecase functions to `js-utils`. How does that sound?

File: client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx
Comment (tyxla) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: Yep, sounds like that should be addressed separately if we decide to go that way.

File: client/lib/wpcom-undocumented/lib/undocumented.js
Comment (jsnajdr) [inline client/lib/wpcom-undocumented/lib/undocumented.js:287]: There's one suspicious thing about the `/me/domain-contact-information/validate` endpoint: it has version 1.2, but we're using the v1.1 one.

If I understand the PHP implementation correctly, the difference is the format of nested validation error messages in the response.

the new v1.2:
```
messages: {
  "extra.ca.cira_agreement_accepted": [
    "Please review and accept the CIRA Registrant Agreement."
  ]
}
```
in the old v1.1 only the last part of the name is sent, possibly ambiguous:
```
messages: {
  "cira_agreement_accepted": [
    "Please review and accept the CIRA Registrant Agreement."
  ]
}
```

@sirbrillig Do you know if the REST response really has the shape we expect? The `composite-checkout` package has some tests for these `extra` domain fields.

**Review by jsnajdr — APPROVED** [review]
Looks good to me :+1:

File: client/lib/wpcom-undocumented/lib/undocumented.js
Comment (tyxla) [reply client/lib/wpcom-undocumented/lib/undocumented.js:287]: We are still using the 1.2 endpoint FWIW: https://github.com/Automattic/wp-calypso/blob/b6894d955505e48f0113dc8e655aa3aba3ba749f/client/my-sites/checkout/composite-checkout/lib/contact-validation.tsx#L310

I'll proceed and ship this as the previous version of the code was using the v1.1 of the endpoint, but I'm happy to revisit if there's the need for that.

File: client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx
Comment (tyxla) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: Moved the methods to `js-utils` in https://github.com/Automattic/wp-calypso/pull/58110

File: client/lib/wpcom-undocumented/lib/undocumented.js
Comment (sirbrillig) [reply client/lib/wpcom-undocumented/lib/undocumented.js:287]: > If I understand the PHP implementation correctly, the difference is the format of nested validation error messages in the response.

You're correct about that. When we remade checkout, we wanted a flat version of the error messages to make it easier to iterate over them and assign them to our new managed data structure, so we created v1.2 of the endpoint for that purpose.

These forms predate the checkout changes and still use the formStateController mechanism for errors, so (I think) they require the old version of the endpoint.

**Pull request merged**
